#!/bin/bash -l

# EXAMPLE SCRIPT ON HOW TO RUN SIMULATIONS AND MAKE SEQUENCES
# REQUIREMENTS: PYTHON, SEQ-GEN, MOTHUR
# MAKE SURE YOU HAVE BOTH THE PIPELINE AND SIM DIRECTORIES IN YOUR SHELL PATH

# YOU SHOULD RUN THIS AFTER YOU HAVE GENERATED EMPIRICAL DATA USING THE PIPELINE

#### VARIABLES ####
# NUMBER OF SIMULATIONS TO RUN
N_SIMULATIONS=1000

# DISTRIBUTION OF GENE LENGTHS
LENGTH_FILE=lengths.txt

# DISTRIBUTION OF NUMBER OF SEQUENCES PER GENE FAMILY
DIST_FILE=dist.txt

# DISTRIBUTION OF NUMBER OF SEQUENCES PER GENE FAMILY
NSEQS_FILE=nseqs.txt

# NUMBER OF PROCESSORS TO USE
N_PROC=1

# ENTER WHERE YOUR EMPIRICAL GENE FAMILY DATA IS STORED
# YOU SHOULD CHANGE THIS!!!
DATA_FOLDER="data"

# WHERE TO STORE SIMULATED NEWICK TREES
TREE_SIM_FOLDER="treesim"

#### SIMULATION PIPELINE ####

# GET NUCLEOTIDE FREQUENCIES
NT_FREQ=`get_nt_freq.py $DATA_FOLDER/*.aligned.fa.filter.fasta`

# GET GENE LENGTHS
parse_lengths.py -ufl temp_$LENGTH_FILE $DATA_FOLDER/*.aligned.fa.filter.fasta
cut -f 2 temp_$LENGTH_FILE > $LENGTH_FILE
rm temp_$LENGTH_FILE

# GET MAX DISTANCES & NUMBER OF SEQUENCES PER FAMILY
for FILE in $DATA_FOLDER/*.aligned.fa.filter.phylip.fn.list; do
  cut -f 1 $FILE | tail -n 1 >> $DIST_FILE
  cut -f 2 $FILE | head -n 1 >> $NSEQS_FILE
done

# RUN TREE SIMLUATIONS
if [ ! -d $TREE_SIM_FOLDER ]; then
  mkdir $TREE_SIM_FOLDER
fi
cd $TREE_SIM_FOLDER
tree_sim.py -n $N_SIMULATIONS --taxafile=../$NSEQS_FILE -o nwk -i --threads=$N_PROC

# MAKE SIMULATED SEQUENCES
run_seq_gen.py -f "$NT_FREQ" --scalefile=$DIST_FILE --lengthfile=$LENGTH_FILE *.nwk

# NOW ANALYZE THEM WITH MOTHUR TO GENERATE DATA TO COMPARE TO EMPIRICAL FAMILIES
run_mothur_pipeline.py *.fa
cd ..

# NOW PARSE THE MOTHUR RESULTS
parse_mothur.py -o $TREE_SIM_FOLDER.txt $TREE_SIM_FOLDER/*.phylip.fn.list

# NOW YOU CAN USE $TREE_SIM_FOLDER.txt AS YOUR SIMULATED NEUTRAL DATA TO COMPARE TO REAL EMPIRICAL DATA!
